icu4c/source/tools/genrb/filterrb.cpp

```diff
@@ -0,0 +1,161 @@
+    if (value.pieces().size() == 0) {
+        out << "/";
+    } else for (auto& key : value.pieces()) {
+        out << "/" << key;
+    }
+    return out;
+}
+
+
+void SimpleRuleBasedPathFilter::addRule(const std::string& ruleLine, UErrorCode& status) {
+    if (ruleLine.length() < 1) {
+        std::cerr << "genrb error: filter rule too short: " << ruleLine << std::endl;
+        status = U_PARSE_ERROR;
+        return;
+    }
+    bool inclusionRule = false;
+    if (ruleLine[0] == '+') {
+        inclusionRule = true;
+    } else if (ruleLine[0] != '-') {
+        std::cerr << "genrb error: lines must start with + or -: " << ruleLine << std::endl;
+        status = U_PARSE_ERROR;
+        return;
+    }
+    ResKeyPath path(ruleLine.substr(1), status);
+    addRule(path, inclusionRule, status);
+}
+
+void SimpleRuleBasedPathFilter::addRule(const ResKeyPath& path, bool inclusionRule, UErrorCode& status) {
+    if (U_FAILURE(status)) {
+        return;
+    }
+    Tree* node = &fRoot;
+    for (auto& key : path.pieces()) {
+        // note: operator[] auto-constructs default values
+        node = &node->fChildren[key];
+    }
+    if (node->fIncluded != PARTIAL) {
+        std::cerr << "genrb error: duplicate paths in filter file" << std::endl;
+        status = U_PARSE_ERROR;
+        return;
+    }
+    node->fIncluded = inclusionRule ? INCLUDE : EXCLUDE;
+}
+
+PathFilter::EInclusion SimpleRuleBasedPathFilter::match(const ResKeyPath& path) const {
+    const Tree* node = &fRoot;
+
+    // defaultResult "bubbles up" the nearest "definite" inclusion/exclusion rule
+    EInclusion defaultResult = INCLUDE;
+    if (node->fIncluded != PARTIAL) {
+        // rules handled here: "+/" and "-/"
+        defaultResult = node->fIncluded;
+    }
+
+    // isLeaf is whether the filter tree can provide no additional information
+    // even if additional subpaths are added to the given key
+    bool isLeaf = false;
+
+    for (auto& key : path.pieces()) {
+        auto child = node->fChildren.find(key);
+        // Leaf case 1: input path descends outside the filter tree
+        if (child == node->fChildren.end()) {
+            isLeaf = true;
+            break;
+        }
+        node = &child->second;
+        if (node->fIncluded != PARTIAL) {
+            defaultResult = node->fIncluded;
+        }
+    }
+
+    // Leaf case 2: input path exactly matches a filter leaf
+    if (node->fChildren.size() == 0) {
+        isLeaf = true;
+    }
+
+    // Always return PARTIAL if we are not at a leaf
+    if (!isLeaf) {
+        return PARTIAL;
+    }
+
+    // If leaf node is PARTIAL, return the default
+    if (node->fIncluded == PARTIAL) {
+        return defaultResult;
+    }
+
+    return node->fIncluded;
+}
+
+void SimpleRuleBasedPathFilter::Tree::print(std::ostream& out, int32_t indent) const {
+    for (int32_t i=0; i<indent; i++) out << "\t";
+    out << "included: " << kEInclusionNames[fIncluded] << std::endl;
```

Comment:
Looks like it doesn't print the fIncluded on the same line as the name of the node?

Reply:
It prints it in a sort-of JSON-like syntax.  I added an example output to the class docs.  Users are not meant to see this; I think this print function will become dead code except during serious debugging.  Do you think we should make the output syntax a little cleaner, like you suggested in your previous comment, and show it on verbose mode?

Reply:
I have some ideas about the output, but let's look at that later.